Added error type checking, added session expiry check, updated tests

morses-code opened the pull request:
- Added error type checking
- Renamed result to sess (represents a session)
- Renamed resultJSON to sessJSON
- Used error types in test cases where required
- Added test for expired sessions
- Added expiry check in the cache
- Created errors package 
- Added new field for Sessions, lastAccessed
- Updated swagger spec with new GET method

morses-code committed:
Added error type checking, added session expiry check, updated tests

- Added error type checking
- Renamed result to sess (represents a session)
- Renamed resultJSON to sessJSON
- Used error types in test cases where required
- Added test for expired sessions
- Added expiry check in the cache
- Created errors package 
- Added new field for Sessions, lastAccessed
- Updated swagger spec with new GET method

nshumoogum commented on api/handlers.go:
```diff
@@ -110,3 +111,13 @@ func writeErrorResponse(ctx context.Context, w http.ResponseWriter, msg string,
+
+func getErrorStatus(err error) int {
+	var status int
+	if errors.Is(err, SessionNotFound) {
```
If both statements fail, the status code will be set to 0 (default integer value) this is equivalent to 500 when the http response is interpreted. Which might be what we want but I would suggest to be more explicit, by using something akin to this:
```
...
	switch {
	case errors.Is(err, SessionNotFound):
		status = http.StatusNotFound
	case errors.Is(err, SessionExpired):
		status = http.StatusNotFound
	default:
		status = http.StatusInternalServerError
	}
...
```

Firstly it explicitly sets default status 500 (internal server error) if a case does not match and simplifies future extension of other errors created in the app.

morses-code replied:
This is exactly what I was trying to do in the first place... I was a bit too Java minded about it though by the looks.

nshumoogum commented on errors/errors.go:
```diff
@@ -0,0 +1,10 @@
+	SessionExpired = errors.New("session has expired")
+)
+
+
```
One too many empty lines

morses-code replied:
👍 removed now.

nshumoogum commented on api/handlers_test.go:
```diff
@@ -255,6 +256,28 @@ func TestGetByIDSessionHandlerFunc(t *testing.T) {
+	Convey("Given a valid request", t, func() {
+		mockCache := &CacheMock{
+			GetByIDFunc: func(ID string) (*session.Session, error) {
+				return nil, SessionExpired
```
I would add a test for when the error returned from GetByID is an unknown error then a 500 internal service error is expected in the response status code.

I understand the cache package only returns two types of errors but if the cache was extended then there wouldn't be a case to handle any new errors and the status will be 0 (not a valid status code even though http interprets it as 500). 

The handler shouldn't have to know this level of detail and we should have a test that proves it can handle the scenario. Also it will force you to think about my previous comment as this test will show you are setting a response code of 0 instead of 500.

for example:
```
...
	mockCache := &CacheMock{
		GetByIDFunc: func(ID string) (*session.Session, error) {
			return nil, errors.New("unexpected error")
		},
	}
...
```

morses-code replied:
Added the new test

morses-code committed:
Addressed PR comments

- Updated `getErrorStatus()` to a `switch` statement
- Added new test to check any other errors return a 500 code
- Removed extra line spaces from errors

morses-code committed:
Update LastAccessed to refresh the TTL
